homeassistant/components/sensor/fronius.py

```diff
@@ -0,0 +1,158 @@
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_TYPE): vol.In(SENSOR_TYPES),
+    vol.Optional(CONF_SCOPE, default=DEFAULT_SCOPE):
+        vol.All(cv.ensure_list, [vol.In(SCOPE_TYPES)]),
+    vol.Optional(CONF_DEVICEID): cv.positive_int,
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up of Fronius platform."""
+    _LOGGER.debug(config)
+    import pyfronius
+
+    session = async_get_clientsession(hass)
+    fronius = pyfronius.Fronius(session, config[CONF_HOST])
+
+    name = "fronius_{}_{}".format(config[CONF_TYPE], config[CONF_HOST])
+    if CONF_DEVICEID in config.keys():
+        device = config[CONF_DEVICEID]
+        name = name + "_{}".format(device)
+    else:
+        device = DEFAULT_DEVICE
+
+    sensor = FroniusSensor(
+        fronius, name, config[CONF_TYPE], config[CONF_SCOPE], device)
+
+    async_add_devices([sensor])
+
+    @asyncio.coroutine
+    def async_fronius(event):
+        """Update all the fronius sensors."""
+        try:
+            yield from sensor.async_update()
+        except Exception:
+            _LOGGER.error("Update of sensor data failed.")
+
+    interval = config.get(CONF_SCAN_INTERVAL) or timedelta(seconds=10)
+    async_track_time_interval(hass, async_fronius, interval)
+
+
+class FroniusSensor(Entity):
+    """The Fronius sensor implementation."""
+
+    def __init__(self, data, name, device_type, scope, device):
+        """Initialize the sensor."""
+        self.data = data
+        self._name = name
+        self._type = device_type
+        self._device = device
+        self._scope = scope
+        self._state = None
+        self._attributes = {}
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the current state."""
+        return self._state
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Retrieve and update latest state."""
+        _LOGGER.debug("Update {}".format(self.name))
+
+        values = {}
+
+        try:
+            values = yield from self._update()
+        except ServerDisconnectedError:
+            _LOGGER.error("Sensor data cannot be updated: connection error.")
+        except TimeoutError:
+            _LOGGER.error("Sensor data cannot be updated: timeout.")
+
+        _LOGGER.debug(values)
+
+        if values:
+            self._state = values['status']['Code']
+            self._attributes = self._get_attributes(values)
+
+    @asyncio.coroutine
+    def _update(self):
+        """Get the values for the current state."""
+        if self._type == TYPE_INVERTER:
+            if self._scope == SCOPE_SYSTEM:
+                return self.data.current_system_inverter_data()
```

Comment:
Is this doing I/O ?

Reply:
This triggers an HTTP request, so it is I/O via network.

Reply:
This function is a coroutine. You are not allowed to do I/O inside a coroutine.

Reply:
Remove `asyncio.coroutine` and name your update method `update` instead of `async_update`

Reply:
This call returns a coroutine object according to the source in pyfronius.